### PR TITLE
Add `Client.restart_workers` method

### PR DESCRIPTION
**Pull request by jrbourbeau**
Sometimes users want to restart individual / a subset of workers in their cluster without restarting their entire cluster (with `client.restart()`). This PR adds a `Client.restart_workers` method for this (similar to the existing `Client.retire_workers` method we provide).  

Noting that there is a relatively high level on engagement on https://github.com/dask/distributed/issues/1823. 

Closes https://github.com/dask/distributed/issues/1823

### CODE REVIEW

**Comment by jakirkham** [comment]
Neat! This need comes up a lot.

cc @pentschev @quasiben (who may find this of interest)

**Comment by jrbourbeau** [comment]
Ah, grand. I'd be curious to hear why this comes up in a RAPIDS context -- at least I assume it's RAPIDS-based based on the Ben / Peter pings : )

File: distributed/client.py [hunk 1]
```diff
@@ -3455,6 +3455,44 @@ def restart(self, timeout=no_default, wait_for_workers=True):
...
+
+        >>> workers = client.scheduler_info()['workers']
+
+        From that list you may want to select some workers to restart
+
+        >>> client.restart_workers(workers=['tcp://address:port', ...])
+
+        See Also
+        --------
+        Client.restart
```
Comment (fjetter) [inline distributed/client.py:3484]: I suggest to cross reference this new method in Client.restart as well. I would also appreciate a sentence about the differences between the two methods

File: distributed/client.py [hunk 2]
```diff
@@ -3455,6 +3455,44 @@ def restart(self, timeout=no_default, wait_for_workers=True):
...
+        """
+        info = self.scheduler_info()
+        for worker in workers:
+            if info["workers"][worker]["nanny"] is None:
+                raise ValueError(
+                    f"Restarting workers requires a nanny to be used. Worker {worker} has type {info['workers'][worker]['type']}."
+                )
+        return self.sync(
+            self.scheduler.broadcast, msg={"op": "restart"}, workers=workers, nanny=True
+        )
```
Comment (fjetter) [inline distributed/client.py:3494]: `Nanny.restart` can return `timed out` which should be handled here.

I don't thin we should return the output of the broadcast. I doubt this is useful / a good user facing API

File: distributed/client.py [hunk 3]
```diff
@@ -3455,6 +3455,44 @@ def restart(self, timeout=no_default, wait_for_workers=True):
             self._restart, timeout=timeout, wait_for_workers=wait_for_workers
         )
 
+    def restart_workers(self, workers: list[str]):
```
Comment (fjetter) [inline distributed/client.py:3458]: `Nanny.restart` is taking a timeout argument. We should allow this to be provided here

File: distributed/client.py [hunk 4]
```diff
@@ -3455,6 +3455,44 @@ def restart(self, timeout=no_default, wait_for_workers=True):
...
+        See Also
+        --------
+        Client.restart
+        """
+        info = self.scheduler_info()
+        for worker in workers:
+            if info["workers"][worker]["nanny"] is None:
+                raise ValueError(
+                    f"Restarting workers requires a nanny to be used. Worker {worker} has type {info['workers'][worker]['type']}."
+                )
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/client.py:3491]: What happens if you do not handle this error here? I would expect broadcast to raise if we are selecting nannies but there are None

**Review by fjetter — CHANGES_REQUESTED** [review]

**Review by jacobtomlinson — COMMENTED** [review]
This looks really useful!

**Comment by pentschev** [comment]
Thanks @jrbourbeau for the work on this and @jakirkham for the ping. This is something that I remember coming up a couple of times, mostly from people wanting to have a CPU worker that eventually turns into a GPU worker or vice-versa.

Jacob, you also had some use for that, didn't you? I can't remember what was the exact use case now though.

**Comment by jacobtomlinson** [comment]
> Jacob, you also had some use for that, didn't you? I can't remember what was the exact use case now though.

Kinda, I have some plans around restarting the worker with different config options which is adjacent to this change. I'm interested in something more along the lines of re-deploying the cluster but without losing HPC/Cloud resources that have already been allocated.

See the [dask-agent](https://github.com/jacobtomlinson/dask-agent) experimental repo I was playing with a while ago.

File: distributed/client.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jrbourbeau) [reply distributed/client.py:3491]: Today I think it just ignores non-`Nanny` workers when `nanny=True`

https://github.com/dask/distributed/blob/ec00cd590fbead781fa6f3960fd6f90c6155166e/distributed/scheduler.py#L5864-L5866

In ^ that snippet `address=None` for non-`Nanny` workers

**Comment by jrbourbeau** [comment]
Thanks for the review @fjetter -- just pushed a commit that should handle your suggestions

**Comment by jakirkham** [comment]
There are also times where workers become unusable or need to be refreshed. Currently we tell people to restart the full cluster, but maybe a less drastic option (like this one) would be useful.

**Review by hendrikmakait — COMMENTED** [review]
Is there a specific reason we do not reuse the restarting logic within `Scheduler.restart` and instead implement `Client.restart_worker` with (what seem to be) slightly different post-conditions? I'd rather have the scheduler be responsible for restarting workers and avoid two functions that might drift even further apart in semantics in the future.

_Restarting workers within `Scheduler.restart`_
https://github.com/dask/distributed/blob/6afce9cce6ab402fa44736dc57b39fbc355b4c11/distributed/scheduler.py#L5722-L5804

_Post conditions on `Scheduler.restart`_
https://github.com/dask/distributed/blob/6afce9cce6ab402fa44736dc57b39fbc355b4c11/distributed/scheduler.py#L5678-L5685?

**Comment by jrbourbeau** [comment]
It looks like `Scheduler.restart` restarts all workers and clears all the local state on the scheduler (this is where all the heavy lifting for `Client.restart` is done). This PR is for just restarting a specified set of workers without clearing the scheduler's state. Does that help clarify things?

**Comment by hendrikmakait** [comment]
> Is there a specific reason we do not reuse the restarting logic within `Scheduler.restart` and instead implement `Client.restart_worker` with (what seem to be) slightly different post-conditions? 

I was referring to the snippet from `Scheduler.restart` linked in the previous post which is concerned with restarting nannies (or dropping non-nanny workers). My idea was that this could be factored out of `Scheduler.restart` into a method that could also be called by `Client.restart_workers`, which would enable code reuse and ensure similar (configurable) semantics.

**Comment by hendrikmakait** [comment]
One conceptual issue I see with this PR is that tasks running on restarting workers will fail (or increment their `suspicious` count). At the very least, this should be documented in the docstring. From my perspective, this operation should probably be considered "safe", i.e., restarting workers should not impact the retry-count of tasks currently running on them. We could leave this for a future PR that makes the logic safe and release `Client.restart_workers` with an appropriate warning in the docstring for users that know what they are doing.

**Reproducer**
```python3

@gen_cluster(
    client=True,
    Worker=Nanny,
    nthreads=[("", 1)],
    config={"distributed.scheduler.allowed-failures": 0},
)
async def test_restart_workers_fails_executing_task(c, s, a):
    ev_start = Event()
    ev_block = Event()

    def clog(ev_start, ev_block):
        ev_start.set()
        ev_block.wait()

    fut = c.submit(
        clog,
        ev_start=ev_start,
        ev_block=ev_block,
        key="wait",
    )
    ev_start.wait()
    await c.restart_workers(workers=[a.worker_address])
    assert await fut.result()
```

**Comment by hendrikmakait** [comment]
@jrbourbeau: In #7154, I have started working on underlying changes to `Client.restart` that will enable us to make `Client.restart` and `Client.restart_workers` behave more similar. As suggestions for this PR, I would document that restarting workers is not safe wrt. to the task retries and leave the rest as is for now. In #7154 I plan to expose a scheduler-side call that can be used by `Client.restart_worker` which should take care of my objections.

**Comment by jrbourbeau** [comment]
Thanks @hendrikmakait -- I've included a note about this in the `Client.restart_workers` docstring. Let me know what you think. Otherwise, I think this PR should be good to go

File: distributed/client.py [hunk 5]
```diff
@@ -3432,29 +3432,98 @@ def restart(self, timeout=no_default, wait_for_workers=True):
...
+            Number of seconds to wait
+
+        Notes
+        -----
+        This method differs from :meth:`Client.restart` in that this method
+        simply restarts the specified set of workers, while ``Client.restart``
+        will restart all workers and also reset local state on the cluster
+        (e.g. all keys are released).
+
+        Additionally, this method makes no safety guarantees for tasks that are
```
Comment (hendrikmakait) [inline distributed/client.py:3497]:
```suggestion
        Additionally, this method does not gracefully handle tasks that are
```

**Review by hendrikmakait — APPROVED** [review]
LGTM, thanks @jrbourbeau. It looks like all the points from @fjetter are addressed as well.

**Comment by jrbourbeau** [comment]
Thanks all for the feedback and @fjetter @hendrikmakait for reviewing! Merging this one in -- CI failures are unrelated

**Pull request merged**